cogboard-app/src/main/kotlin/com/cognifide/cogboard/widget/type/ZabbixWidget.kt

```diff
@@ -0,0 +1,128 @@
+package com.cognifide.cogboard.widget.type
+
+import com.cognifide.cogboard.CogboardConstants
+import com.cognifide.cogboard.widget.AsyncWidget
+import io.vertx.core.Vertx
+import io.vertx.core.json.JsonArray
+import io.vertx.core.json.JsonObject
+import io.vertx.core.logging.Logger
+import io.vertx.core.logging.LoggerFactory
+
+class ZabbixWidget(vertx: Vertx, config: JsonObject) : AsyncWidget(vertx, config) {
+
+    private val selectedMetric: String = config.getString(METRIC, "")
+    private val host: String = config.getString(HOST, "")
+
+    val logger: Logger
+        get() = LoggerFactory.getLogger(ZabbixWidget::class.java)
+
+    override fun updateState() {
+        if (authorizationToken.containsKey(publicUrl)) {
+            fetchItemData()
+        } else {
+            fetchAuthorizationToken()
+        }
+    }
+
+    private fun fetchAuthorizationToken() {
+        val body = attachCredentialsToBody()
+        httpPost(publicUrl, body)
+    }
+
+    private fun attachCredentialsToBody(): JsonObject {
+        val credentials = JsonObject()
+                .put(CogboardConstants.PROP_USER, user)
+                .put(CogboardConstants.PROP_PASSWORD, password)
+        return prepareRequestBody(USER_LOGIN, credentials)
+    }
+
+    override fun handleResponse(responseBody: JsonObject) {
+        val body = responseBody.getString("body")
+        val value = JsonObject(body)
+        if (value.getValue(RESULT).toString().contains("key_")) {
+            send(JsonObject().put("lastValue", extractLastValue(value)))
+        } else if (authorizationToken.containsKey(publicUrl) && !value.getString(RESULT).contains(selectedMetric)) {
+            fetchItemData()
+        } else if (JsonObject(body).containsKey(ERROR)) {
+            sendAuthorizationError(JsonObject(body))
+            logger.error(responseBody.getValue(ERROR))
+        } else {
+            saveToken(value.getString(RESULT))
+            fetchItemData()
+        }
+    }
+
+    private fun extractLastValue(value: JsonObject) =
```

Comment:
make an extension function out of this:
https://kotlinlang.org/docs/reference/extensions.html

Reply:
done